flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/AbstractArrowPythonAggregateFunctionOperator.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.flink.table.runtime.operators.python.AbstractStatelessFunctionOperator;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.util.Preconditions;
+
+/**
+ * The Abstract class of Arrow Aggregate Operator for Pandas {@link AggregateFunction}.
+ */
+@Internal
+public abstract class AbstractArrowPythonAggregateFunctionOperator
+	extends AbstractStatelessFunctionOperator<RowData, RowData, RowData> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final String SCHEMA_ARROW_CODER_URN = "flink:coder:schema:arrow:v1";
+
+	private static final String PANDAS_AGGREGATE_FUNCTION_URN = "flink:transform:pandas_aggregate_function:v1";
+
+	/**
+	 * The Pandas {@link AggregateFunction}s to be executed.
+	 */
+	private final PythonFunctionInfo[] pandasAggFunctions;
+
+	protected final int[] groupingSet;
+
+	protected transient ArrowSerializer<RowData> arrowSerializer;
+
+	/**
+	 * The collector used to collect records.
+	 */
+	protected transient StreamRecordRowDataWrappingCollector rowDataWrapper;
+
+	/**
+	 * The JoinedRowData reused holding the execution result.
+	 */
+	protected transient JoinedRowData reuseJoinedRow;
+
+	/**
+	 * The current number of elements to be included in an arrow batch.
+	 */
+	protected transient int currentBatchCount;
+
+	/**
+	 * The Projection which projects the udaf input fields from the input row.
+	 */
+	private transient Projection<RowData, BinaryRowData> udafInputProjection;
+
+	public AbstractArrowPythonAggregateFunctionOperator(
+		Configuration config,
+		PythonFunctionInfo[] pandasAggFunctions,
+		RowType inputType,
+		RowType outputType,
+		int[] groupingSet,
+		int[] udafInputOffsets) {
+		super(config, inputType, outputType, udafInputOffsets);
+		this.pandasAggFunctions = Preconditions.checkNotNull(pandasAggFunctions);
+		this.groupingSet = Preconditions.checkNotNull(groupingSet);
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		rowDataWrapper = new StreamRecordRowDataWrappingCollector(output);
+		reuseJoinedRow = new JoinedRowData();
+
+		udafInputProjection = createUdafInputProjection();
+		arrowSerializer = new RowDataArrowSerializer(userDefinedFunctionInputType, userDefinedFunctionOutputType);
+		arrowSerializer.open(bais, baos);
+		currentBatchCount = 0;
+	}
+
+	@Override
+	public void dispose() throws Exception {
+		super.dispose();
+		arrowSerializer.close();
+	}
+
+	@Override
+	public void processElement(StreamRecord<RowData> element) throws Exception {
+		RowData value = element.getValue();
+		bufferInput(value);
+		processElementInternal(value);
+		emitResults();
+	}
+
+	@Override
+	protected void checkInvokeFinishBundleByCount() throws Exception {
+		elementCount += currentBatchCount;
+		currentBatchCount = 0;
+		if (elementCount >= maxBundleSize) {
+			invokeFinishBundle();
+		}
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
```

Comment:
This implementation is very similar to the implementation in the super class. It seems that we add this method just because **elementCount** was computed in **checkInvokeFinishBundleByCount**. If so, we just need to improve this, then we could avoid the duplicate code.

Reply:
Yes. Make sense. I will extract the logic about deciding whether there is buffered unprocessed data into a method so that we can avoid overriding `processWatermark` method